### PR TITLE
Instead of allocating a WASM instance for each fabric, just allocate one

### PR DESCRIPTION
**Pull request by geralddejong**
It was quite a day of coding!  Good news is that it performs better, because instantiating was causing delays.

### CODE REVIEW

File: client/src/app.tsx [hunk 1]
```diff
@@ -99,24 +100,36 @@ function selectSpot(spot?: Spot) {
...
-                fabric.createSeed(location.x, location.z)
-                return new Gotchi(fabric, genome)
+        this.fabricKernel = createFabricKernel(props.fabricExports, MAX_POPULATION, INITIAL_JOINT_COUNT)
+        this.instanceUsed = this.fabricKernel.instance.map(() => false)
+        const createGotchiAt = (location: Vector3, genome: Genome): Gotchi => {
+            const freeIndex = this.instanceUsed.indexOf(false)
+            if (freeIndex < 0) {
+                throw new Error("No free fabrics!")
+            }
+            // console.log("Allocating fabric", freeIndex)
```
Comment (mitchelljustin) [inline client/src/app.tsx]: Remove

File: client/src/app.tsx [hunk 2]
```diff
@@ -99,24 +100,36 @@ function selectSpot(spot?: Spot) {
...
+            this.instanceUsed[freeIndex] = true
+            const exports = this.fabricKernel.instance[freeIndex]
+            exports.flushFaces()
+            const fabric = new Fabric(exports, freeIndex).createSeed(location.x, location.z)
+            fabric.iterate(0)
+            return new Gotchi(fabric, genome, (index: number) => {
+                // console.log("Freeing fabric", index)
```
Comment (mitchelljustin) [inline client/src/app.tsx]: Remove

File: client/src/app.tsx [hunk 3]
```diff
@@ -99,24 +100,36 @@ function selectSpot(spot?: Spot) {
...
+            if (freeIndex < 0) {
+                throw new Error("No free fabrics!")
+            }
+            // console.log("Allocating fabric", freeIndex)
+            this.instanceUsed[freeIndex] = true
+            const exports = this.fabricKernel.instance[freeIndex]
+            exports.flushFaces()
+            const fabric = new Fabric(exports, freeIndex).createSeed(location.x, location.z)
+            fabric.iterate(0)
+            return new Gotchi(fabric, genome, (index: number) => {
```
Comment (mitchelljustin) [inline client/src/app.tsx]: What's this constructor arg? You pass in a func which takes an index? Why?

Comment (geralddejong) [reply client/src/app.tsx]: The index is its "address" in memory.  0..(N-1)

Comment (mitchelljustin) [reply client/src/app.tsx]: Is there no non-async way to do this?

File: tslint.json [hunk 4]
```diff
@@ -55,6 +55,8 @@
       true,
       2
     ],
+    "prefer-const": false, // for the assemblyscript
+    "no-angle-bracket-type-assertion": false, // for the assemblyscript
```
Comment (mitchelljustin) [inline tslint.json]: Lol, i don't think // comments work in JSON files?

File: client/src/gotchi/gotchi.ts [hunk 5]
```diff
@@ -40,9 +45,9 @@ export class Gotchi {
         return Math.sqrt(xx * xx + zz * zz)
     }
 
-    public withNewBody(fabric: Fabric): Gotchi {
-        return new Gotchi(fabric, this.genome)
-    }
+    // public withNewBody(fabric: Fabric): Gotchi {
+    //     return new Gotchi(fabric, this.genome)
+    // }
```
Comment (mitchelljustin) [inline client/src/gotchi/gotchi.ts]: Delete

File: client/src/body/fabric-kernel.ts [hunk 6]
```diff
@@ -11,13 +11,74 @@ export const vectorFromFloatArray = (array: Float32Array, index: number, vector?
...
+        dimensions: IFabricDimensions,
+    ) {
+        const fabricBytes = exports.init(dimensions.jointCountMax, dimensions.intervalCountMax, dimensions.faceCountMax, dimensions.instanceMax)
+        this.offsets = createOffsets(dimensions.faceCountMax, fabricBytes)
+        const byteLength = exports.memory.buffer.byteLength
+        if (byteLength === 0) {
+            throw new Error(`Zero byte length! ${this.offsets.fabricBytes}`)
+        } else {
+            console.log(`Got ${byteLength} bytes`)
+        }
```
Comment (mitchelljustin) [inline client/src/body/fabric-kernel.ts:67]: Can you remove these logs now that you've fixed the problem?

File: client/src/app.tsx [hunk 7]
```diff
@@ -99,25 +100,33 @@ function selectSpot(spot?: Spot) {
...
-                return new Gotchi(fabric, genome)
-            })
+        this.fabricKernel = createFabricKernel(props.fabricExports, MAX_POPULATION, INITIAL_JOINT_COUNT)
+        this.instanceUsed = this.fabricKernel.instance.map(() => false)
+        const createGotchiAt = (location: Vector3, genome: Genome): Gotchi => {
+            const freeIndex = this.instanceUsed.indexOf(false)
+            if (freeIndex < 0) {
+                throw new Error("No free fabrics!")
+            }
+            this.instanceUsed[freeIndex] = true
```
Comment (mitchelljustin) [inline client/src/app.tsx:124]: Pretty simple memory management :)

**Review by mitchelljustin — APPROVED** [review]

**Pull request merged**